flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/BulkFormatMapping.java

```diff
@@ -0,0 +1,133 @@
+        this.bulkFormat = bulkFormat;
+    }
+
+    @Nullable
+    public int[] getIndexMapping() {
+        return indexMapping;
+    }
+
+    public BulkFormat<RowData, FileSourceSplit> getReaderFactory() {
+        return bulkFormat;
+    }
+
+    public static BulkFormatMappingBuilder newBuilder(
+            FileFormat fileFormat,
+            KeyValueFieldsExtractor extractor,
+            int[][] keyProjection,
+            int[][] valueProjection,
+            int[][] projection,
+            @Nullable List<Predicate> filters) {
+        return new BulkFormatMappingBuilder(
+                fileFormat, extractor, keyProjection, valueProjection, projection, filters);
+    }
+
+    /** Builder to build {@link BulkFormatMapping}. */
+    public static class BulkFormatMappingBuilder {
+        private final FileFormat fileFormat;
+        private final KeyValueFieldsExtractor extractor;
+        private final int[][] keyProjection;
+        private final int[][] valueProjection;
+        private final int[][] projection;
+        @Nullable private final List<Predicate> filters;
+
+        private BulkFormatMappingBuilder(
+                FileFormat fileFormat,
+                KeyValueFieldsExtractor extractor,
+                int[][] keyProjection,
+                int[][] valueProjection,
+                int[][] projection,
+                @Nullable List<Predicate> filters) {
+            this.fileFormat = fileFormat;
+            this.extractor = extractor;
+            this.keyProjection = keyProjection;
+            this.valueProjection = valueProjection;
+            this.projection = projection;
+            this.filters = filters;
+        }
+
+        public BulkFormatMapping build(TableSchema tableSchema, TableSchema dataSchema) {
+            List<DataField> tableKeyFields = extractor.keyFields(tableSchema);
+            List<DataField> tableValueFields = extractor.valueFields(tableSchema);
+
+            List<DataField> dataKeyFields = extractor.keyFields(dataSchema);
+            List<DataField> dataValueFields = extractor.valueFields(dataSchema);
+            int[][] dataKeyProjection =
+                    SchemaEvolutionUtil.createDataProjection(
+                            tableKeyFields, dataKeyFields, keyProjection);
+            int[][] dataValueProjection =
+                    SchemaEvolutionUtil.createDataProjection(
+                            tableValueFields, dataValueFields, valueProjection);
+
+            RowType keyType = RowDataType.toRowType(false, dataKeyFields);
+            RowType valueType = RowDataType.toRowType(false, dataValueFields);
+            RowType dataRecordType = KeyValue.schema(keyType, valueType);
+            int[][] dataProjection =
+                    KeyValue.project(
+                            dataKeyProjection, dataValueProjection, keyType.getFieldCount());
+
+            int[] indexMapping =
+                    SchemaEvolutionUtil.createIndexMapping(
```

Comment:
I don't quite understand why we do createDataProjection first and then createIndexMapping. Can you give me an example?

Reply:
This process is really complicated, and I'm crazy.
We can add more comments to explain every step.